src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -812,16 +831,30 @@ vector<Visitor::ExpressionPtr> Visitor::visitTypesafe(
   return visitVector(ctx->expression());
 }
 
-// ____________________________________________________________________________________
-Triples Visitor::visitTypesafe(Parser::ConstructTriplesContext* ctx) {
+template <typename Context>
+Triples Visitor::parseTriplesConstruction(Context* ctx,
+                                          Context* (Context::*F)(void)) {
   auto result = visitTypesafe(ctx->triplesSameSubject());
-  if (ctx->constructTriples()) {
-    auto newTriples = visitTypesafe(ctx->constructTriples());
+  Context* subContext = std::invoke(F, ctx);
+  if (subContext) {
+    auto newTriples = visitTypesafe(subContext);
     ad_utility::appendVector(result, std::move(newTriples));
```

Comment:
This needs really good comments as for the what and why and how. I have understood it in the meantime, but at first glance it is confusing.

Reply:
I think, since you only need the two cases, you can write it as:
`if constexpr (std::is_same_v<Context, ConstructTriplesContext>) {
  subContext = ctx->constructTriples() }`
(similar branch for the other case together with a static assert that these are the only two cases, and maybe put this if-else into an immediately invoked lambda, then it gets even more understandable.